### ISSUE TITLE
Do not cache URLs with query strings does nothing

### ISSUE THREAD

**Issue by samuelreichor** [putyourlightson/craft-blitz#729]
### Bug Report

I want to disable caching for pages with query parameters, but it doesn't seem to work as expected.

Here are the settings I've configured (which I believe should be correct):

Query String Caching: Do not cache URLs with query strings
1. Included Query String Parameters: (empty)
2. Included URI Patterns: ^studienangebot$
3. However, URLs like /studienangebot?fTitle=bachelor are still being cached, even though they shouldn't be.

Interestingly, if I change the settings as follows:

Query String Caching: Do not cache URLs with query strings
1. Included Query String Parameters: fTitle
2. Included URI Patterns: ^studienangebot$
3. Then, URLs like /studienangebot?fTitle=bachelor are not cached, which is the expected behavior.

In debug mode, I receive the message:
Page not cached because it does not match an included URI pattern. /studienangebot?fTitle=bachelor

I’ve tried investigating further and it seems like there might be an issue with how CacheRequestService.php is handling this logic, though this is just a guess. You'd certainly have a better understanding of how this should behave.

Here’s what seem strange to me: 
```php
if (Blitz::$plugin->settings->queryStringCaching == SettingsModel::QUERY_STRINGS_DO_NOT_CACHE_URLS) {
    $allowedQueryString = $this->getAllowedQueryString($siteUri->siteId, $siteUri->uri);
    dd($allowedQueryString); // Returns '' with no included query string params, but shouldn't the function return false in this context?
    if ($allowedQueryString) {
        Blitz::$plugin->debug('Page not cached because a query string was provided with the query string caching setting disabled.', [], $url);
        return false;
    }
}
```

#### Application Info

- PHP version: 8.1.29
- Craft edition & version: Pro 4.12.1
- Database driver & version: MariaDB 10.4.34

#### Installed Plugins

- Blitz: 4.x-dev
- Control Panel CSS: 2.6.0
- Custom Login: 1.0.0
- Empty Coalesce: 4.0.0
- Expanded Singles: 2.0.5
- Feed Me: 5.7.0
- Formie: 2.1.29
- Image Resizer: 3.0.12
- Imager X: 4.4.1
- Lockout: 2.0.0
- Mailgun: 3.1.0
- No-Cache: 3.0.2
- PDF Renderer: 1.3.0
- Proper - Clean Frontend Templates: 1.6.1
- Redactor: 3.1.0
- Retour: 4.1.19
- Sentry SDK: 3.0.0
- SEOmatic: 4.1.2
- Site Copy X: 1.1.0
- Smith: 2.0.1
- Spoon: 4.0.4
- Super Table: 3.0.14
- Table Maker: dev-craft-4
- Typed link field: 2.1.5
- Vite: 4.0.10

#### Loaded Modules

- a11y: A11y\A11y
- careerzoneFeed: CareerzoneFeed\CareerzoneFeed
- filteredContent: FilteredContent\FilteredContent
- logaImport: LogaImport\LogaImportModule
- navigation: Navigation\Navigation
- projectUtils: ProjectUtils\ProjectUtils
- pvsVeseCrmImport: PvsVeseCrmImport\PvsVeseCrmImportModule
- searchTools: SearchTools\SearchTools
- textJoiner: TextJoiner\TextJoiner
- themeSetter: ThemeSetter\ThemeSetter
- codeeditor: nystudio107\codeeditor\CodeEditor
- sprig-core: putyourlightson\sprig\Sprig
- blitz-hints: putyourlightson\blitzhints\BlitzHints
- verbb-base: verbb\base\Base

#### Blitz Plugin Settings

```json
{
    "debug": false,
    "hintsEnabled": true,
    "cachingEnabled": true,
    "refreshCacheEnabled": true,
    "refreshMode": 3,
    "includedUriPatterns": {
        "0": {
            "siteId": "1",
            "uriPattern": ""
        },
        "1": {
            "siteId": "2",
            "uriPattern": ""
        },
        "2": {
            "siteId": "1",
            "uriPattern": "^studienangebot$"
        },
        "7": {
            "siteId": "1",
            "uriPattern": "^studienangebot\/bachelor-studiengaenge$"
        },
        "3": {
            "siteId": "1",
            "uriPattern": "^campus-wels$"
        },
        "4": {
            "siteId": "1",
            "uriPattern": "^campus-steyr$"
        },
        "5": {
            "siteId": "1",
            "uriPattern": "^campus-hagenberg$"
        },
        "6": {
            "siteId": "1",
            "uriPattern": "^campus-linz$"
        }
    },
    "excludedUriPatterns": [],
    "cacheStorageType": "putyourlightson\\blitz\\drivers\\storage\\FileStorage",
    "cacheStorageSettings": {
        "folderPath": "@webroot\/cache\/blitz",
        "compressCachedValues": ""
    },
    "cacheStorageTypes": [],
    "cacheGeneratorType": "putyourlightson\\blitz\\drivers\\generators\\HttpGenerator",
    "cacheGeneratorSettings": {
        "concurrency": "3"
    },
    "cacheGeneratorTypes": [],
    "customSiteUris": [],
    "cachePurgerType": "putyourlightson\\blitz\\drivers\\purgers\\DummyPurger",
    "cachePurgerSettings": [],
    "cachePurgerTypes": [],
    "deployerType": "putyourlightson\\blitz\\drivers\\deployers\\DummyDeployer",
    "deployerSettings": [],
    "deployerTypes": [],
    "ssiEnabled": false,
    "ssiTagFormat": "<!--#include virtual=\"{uri}\" -->",
    "detectSsiEnabled": true,
    "esiEnabled": false,
    "queryStringCaching": 0,
    "includedQueryStringParams": "",
    "excludedQueryStringParams": [
        {
            "siteId": "",
            "queryStringParam": "gclid"
        },
        {
            "siteId": "",
            "queryStringParam": "fbclid"
        }
    ],
    "apiKey": "",
    "generatePagesWithQueryStringParams": true,
    "purgeAssetImagesWhenChanged": true,
    "refreshCacheAutomaticallyForGlobals": true,
    "refreshCacheWhenElementMovedInStructure": true,
    "refreshCacheWhenElementSavedUnchanged": false,
    "refreshCacheWhenElementSavedNotLive": false,
    "cacheNonHtmlResponses": false,
    "trackElements": true,
    "trackElementQueries": true,
    "excludedTrackedElementQueryParams": [],
    "cacheElements": true,
    "cacheElementQueries": true,
    "cacheDuration": null,
    "nonCacheableElementTypes": [],
    "sourceIdAttributes": [],
    "liveStatuses": [],
    "integrations": [
        "putyourlightson\\blitz\\drivers\\integrations\\CommerceIntegration",
        "putyourlightson\\blitz\\drivers\\integrations\\SeomaticIntegration"
    ],
    "defaultCacheControlHeader": "no-store",
    "cacheControlHeader": "public, s-maxage=31536000, max-age=0",
    "cacheControlHeaderExpired": "public, s-maxage=5, max-age=0",
    "sendPoweredByHeader": true,
    "outputComments": true,
    "refreshCacheJobPriority": 10,
    "driverJobBatchSize": 100,
    "driverJobPriority": 100,
    "queueJobTtr": 300,
    "maxRetryAttempts": 10,
    "maxUriLength": 255,
    "mutexTimeout": 1,
    "commands": [],
    "injectScriptEvent": "DOMContentLoaded",
    "injectScriptPosition": 3
}
```

#### Recommendations

- ❌ One or more globals exist and refreshCacheAutomaticallyForGlobals is enabled.
- ❌ The Blitz Hints utility is reporting 17 eager-loading opportunities.
- ❌ The blitz/cache/refresh-expired console command has not been executed within the past 24 hours.
- ✅ Blitz is configured not to refresh cached pages when an element is saved but unchanged.
- ✅ Blitz is configured not to refresh cached pages when an element is saved but not live.
- ✅ Image transforms are configured to be generated before page load.
- ✅ The @web alias is explicitly defined.
- ✅ Queue jobs are configured not to run automatically via web requests.
- ✅ The Async Queue plugin is not installed or enabled.

#### Site Tracking [1]

- Tracked Pages: 3
- Tracked Includes: 0
- Tracked Query String Params: 0
- Tracked Elements: 220
	- craft\elements\Entry: 174
	- craft\elements\Asset: 46
- Tracked Element Queries: 13
	- craft\elements\Entry: 13
- Tracked Tags: 11

#### Site Tracking [2]

- Tracked Pages: 0
- Tracked Includes: 0
- Tracked Query String Params: 0
- Tracked Elements: 0
- Tracked Element Queries: 0
- Tracked Tags: 0

#### Site Tracking [3]

- Tracked Pages: 0
- Tracked Includes: 0
- Tracked Query String Params: 0
- Tracked Elements: 0
- Tracked Element Queries: 0
- Tracked Tags: 0

#### Site Tracking [4]

- Tracked Pages: 0
- Tracked Includes: 0
- Tracked Query String Params: 0
- Tracked Elements: 0
- Tracked Element Queries: 0
- Tracked Tags: 0

**Comment by bencroker** [comment]
Thanks for the report. In Discord I sent you https://github.com/putyourlightson/craft-blitz/commit/0da5186d69cca474a0ef0c05faaee8a48566943d, which addresses this. Did you clear the cache and retest?

**Comment by samuelreichor** [comment]
Yes I also deleted vendor and made a fresh new install. I flushed and deleted the cache.

**Comment by bencroker** [comment]
Ok, I'll troubleshoot and let you know what I find.

**Comment by bencroker** [comment]
The latest commit is working for me locally. Please ensure you’re using the latest version of the code, clear and flush the cache, and retest. If the issue persists, describe the steps you took in testing in detail so I can follow them.

**Comment by samuelreichor** [comment]
I tested everything again, but unfortunately, it’s still not working for me.

### Test Procedure:

1. **Double-checked all settings**: I confirmed that all settings are configured correctly.  
   ![image](https://github.com/user-attachments/assets/b2975dae-acb4-454c-b04a-0256def72a4b)  
   ![image](https://github.com/user-attachments/assets/7db05186-f80a-4c2d-aed9-c058ea83b750)

2. **Cleared the cache** using the utilities in Craft CMS.

3. **Flushed the cache** again through Craft’s utilities.

4. **Updated `putyourlightson/craft-blitz`** to this version:
   ```json
   "putyourlightson/craft-blitz": "4.x-dev#0da5186d69cca474a0ef0c05faaee8a48566943d",
   ```

5. **Verified the plugin update** in the `vendor` folder by comparing it with the latest changes in Git – everything appears correct.

6. **Checked caching for `/studienangebot?fCampus=hagenberg&fTitle=bachelor`**: This page is served by blitz.

7. **Checked caching for `/studienangebot`**: This page is served by blitz. (But the cached Page is wrong because of `?fCampus=hagenberg&fTitle=bachelor` from the first hit)

**Comment by samuelreichor** [comment]
(And I tested in`ENVIRONMENT=production`)

**Comment by bencroker** [comment]
> 6. Checked caching for /studienangebot?fCampus=hagenberg&fTitle=bachelor: This page is served by blitz.

“Served by Blitz” doesn’t mean that the page was cached by Blitz. How are you verifying whether the page is cached?

**Comment by samuelreichor** [comment]
Sorry, that was not expressed correctly. I see two comments from Blitz which shows that it is served and cached by blitz

**Comment by bencroker** [comment]
Thanks, I was able to replicate the issue by matching your “Included Query String Parameters”. Can you please retest with the commit https://github.com/putyourlightson/craft-blitz/commit/b65c92e90430951c9b5cf11cfce24c0b45e0a06b?

**Comment by samuelreichor** [comment]
Ah, dang it, now the usual URI include patterns are broken. 🙈

I tested it as before, cleared and flushed the cache, but now all the 'normal' include URI patterns are being ignored. Strangely enough, the homepage is still working fine.

Here's what's showing up in the logs in debug mode:

```
[2024-11-07 14:40:09] Page not cached because it does not match an included URI pattern. [https://fhooe.ddev.site/studienangebot?fCampus=hagenberg&fTitle=bachelor]
[2024-11-07 14:41:27] Page not cached because it does not match an included URI pattern. [https://fhooe.ddev.site/studienangebot]
[2024-11-07 14:41:39] Page not cached because it does not match an included URI pattern. [https://fhooe.ddev.site/campus-hagenberg]
```

**Comment by bencroker** [comment]
Indeed. This should hopefully be the final fix https://github.com/putyourlightson/craft-blitz/commit/c05b92117b57d8b6b70f2358cabdabd6778b35a5.

**Comment by samuelreichor** [comment]
Yessss! that works now, thank you! :)

**Comment by bencroker** [comment]
Great, I’ll be releasing this today or tomorrow. Thanks for reporting this and for sticking with me getting through it.

**Comment by bencroker** [comment]
Released in [4.23.5](https://github.com/putyourlightson/craft-blitz/releases/tag/4.23.5) and [5.9.3](https://github.com/putyourlightson/craft-blitz/releases/tag/5.9.3).
